scripts/dtls.test

```diff
@@ -0,0 +1,156 @@
+#!/bin/bash
+
+# This script can be run with several environment variables set dictating its
+# run. You can set the following to what you like:
+WOLFSSL_ROOT=${WOLFSSL_ROOT:-$(pwd)}
+UDP_PROXY_BIN=${UDP_PROXY_BIN:-"udp_proxy"}
+DTLS_VERSION=${DTLS_VERSION:-"-v4"}
+# Additionally, you can add the following tests by setting it to '1':
+DO_EXTENDED_SERVER_PERMUTATION_TEST=${DO_EXTENDED_SERVER_PERMUTATION_TEST:-0}
+DO_DELAY_TEST=${DO_DELAY_TEST:-0}
+
+# An example use would be: DTLS_VERSION=-v3 scripts/dtls.test
+
+#set -x # enable debug output
+
+cleanup () {
+    echo
+    echo "Cleaning up..."
+    if [ ! -z "$UDP_PROXY_PID" ];then
+        echo "Killing udp_proxy $UDP_PROXY_PID"
+        kill $UDP_PROXY_PID
+    fi
+    if [ ! -z "$SERVER_PID" ];then
+        echo "Killing server $SERVER_PID"
+        kill $SERVER_PID
+    fi
+    if [ ! -z "$TCPDUMP_PID" ];then
+        echo "Killing tcpdump $TCPDUMP_PID"
+        sleep 1
+        kill $TCPDUMP_PID
+    fi
+}
+
+trap cleanup err exit
+
+PROXY_PORT=1234
+SERVER_PORT=4321
+KEY_UPDATE_SIZE=35
+NUM_TESTS_FAILED=0
+NUM_TESTS_RUN=0
+
+if [ "$DTLS_VERSION" = "-v4" ]; then
+    UDP_PROXY_EXTRA_ARGS="-u"
+fi
+
+# $WOLFSSL_ROOT/tests/unit.test tests/test-dtls13.conf
+
+set -o pipefail
+prepend() { # Usage: cmd 2>&1 | prepend "sometext "
+    while read line; do echo "${1}${line}"; done
+}
+
+run_test() { # usage: run_test "<testName>" "<udp-proxy args>" "<server args>" "<client args>"
+    ((NUM_TESTS_RUN++))
+    echo "" | nc -u 127.0.0.1 $SERVER_PORT # This is a marker for the PCAP file
+    echo "$1" | nc -u 127.0.0.1 $SERVER_PORT # This is a marker for the PCAP file
+    echo "" | nc -u 127.0.0.1 $SERVER_PORT # This is a marker for the PCAP file
+    echo -e "\n${1}\n"
+    stdbuf -oL -eL $WOLFSSL_ROOT/examples/server/server -u -p$SERVER_PORT $DTLS_VERSION $3 2>&1 | prepend "[server] " &
+    SERVER_PID=$(($! - 1))
+    stdbuf -oL -eL $UDP_PROXY_BIN -p $PROXY_PORT -s 127.0.0.1:$SERVER_PORT $UDP_PROXY_EXTRA_ARGS $2 2>&1 | prepend "[udp-proxy] " &
+    UDP_PROXY_PID=$(($! - 1))
+    sleep 0.2
+    # Wrap this command in a timeout so that a deadlock won't bring down the entire test
+    timeout -s KILL 5m stdbuf -oL -eL $WOLFSSL_ROOT/examples/client/client -u -p$PROXY_PORT $DTLS_VERSION $4 2>&1 | prepend "[client] "
+    if [ $? != 0 ]; then
+        echo "***Test failed***"
+        ((NUM_TESTS_FAILED++))
+    fi
+    kill $SERVER_PID >&/dev/null # make sure the server is no longer running
+    SERVER_PID=
+    kill $UDP_PROXY_PID
+    UDP_PROXY_PID=
+}
+
+test_dropping_packets () {
+    for i in $(seq 3 11);do
+        run_test "Dropping ${i}th packet" "-d $i" "-Ta" ""
```

Comment:
The `-d` option is not safe for generic testing. It can drop app data and fail the tests. `-d` shouldn't have an affect on 1.2 but it either doesn't work on 1.3 or it can drop real app data since it can't distinguish between handshake msgs and app data.